### PR TITLE
firstboot: mark essential snaps as "Required" in the state

### PR DESCRIPTION
**Pull request by mvo5**
This ensures that we cannot remove the "snapd" or "core18" snaps
on a core18 system. As a side-effect it also sets the kernel and
gadget as Required which is something we want to do anyway.

### CODE REVIEW

**Comment by pedronis** [comment]
the original idea of required was to mark only the required-snaps from the model, and have other implicit knowledge about the others, fair to change this but worth pointing out

**Comment by mvo5** [comment]
@pedronis There is a comment in snapstate.go:canRemove(): https://github.com/snapcore/snapd/blob/master/overlord/snapstate/snapstate.go#L1280 that talks about using Required for os/gadget/kernel. I'm happy to do it differently of course.

**Comment by pedronis** [comment]
@mvo5 now I remember the issue is that we never implemented setting Required for preexisting devices because it was either a patch or code to run each time snapd starts, so we couldn't fully switch, bootable bases and snapd are new though

File: overlord/devicestate/firstboot.go [hunk 1]
```diff
@@ -111,10 +111,9 @@ func populateStateFromSeedImpl(st *state.State) ([]*state.TaskSet, error) {
 		return nil, err
 	}
 
-	var required map[string]bool
+	required := map[string]bool{}
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go]: seems still worth making it at len(reqSnaps) big or len(reqSnaps)+4  (snapd, base, gadget, kernel)

**Pull request merged**